### ISSUE TITLE
Inference length penalty

### ISSUE THREAD

**Issue by FangxuLiu** [awslabs/sockeye#757]
Hello, I found that the length of the translation generated in decoding is short, and I try to modify the length penalty to make the translation longer, but it didn't work. Any help is appreciated!

**Comment by artemsok** [comment]
Hi, could you give more details on what you did? Did you learn to predict length ratio during training, and which inference-time brevity penalty mode did you use?

**Comment by FangxuLiu** [comment]
Hello, I want to train a NMT model based on Transformer.
Training command:
python -m sockeye.train -d $data_path/train_data \
-vs $data_path/test_data/$test_src -vt $data_path/test_data/$test_trg \
--encoder transformer --decoder transformer \
--max-seq-len 256 -o $Model --batch-size 8192 \
--max-updates 100000 --checkpoint-interval 5000
Decoding command:
python -m sockeye.translate --input $input --models $Models --beam-size 4 --device-ids -1 --batch-size 32 --output $input.trans

I found that the length of translation is relatively short, and then I try to modify the parameter "--length-penalty-alpha" or set "--brevity-penalty-type=constant, --brevity-penalty-constant-length-ratio=$alpha", but it didn't work.

**Comment by FangxuLiu** [comment]
> Hi, could you give more details on what you did? Did you learn to predict length ratio during training, and which inference-time brevity penalty mode did you use?

Hello, I want to train a NMT model based on Transformer.
Training command:
python -m sockeye.train -d $data_path/train_data
-vs $data_path/test_data/$test_src -vt $data_path/test_data/$test_trg
--encoder transformer --decoder transformer
--max-seq-len 256 -o $Model --batch-size 8192
--max-updates 100000 --checkpoint-interval 5000
Decoding command:
python -m sockeye.translate --input $input --models $Models --beam-size 4 --device-ids -1 --batch-size 32 --output $input.trans

I found that the length of translation is relatively short, and then I try to modify the parameter "--length-penalty-alpha" or set "--brevity-penalty-type=constant, --brevity-penalty-constant-length-ratio=$alpha", but it didn't work.

**Comment by artemsok** [comment]
```--length-penalty-alpha``` and ```--brevity-penalty-*``` are two different mechanisms of length control. You can play with the former, but in my experience it was hard to fix shortening problems; I also didn't try mixing them at inference. The latter mechanism has ```constant``` mode (as in your example; works with default training) or ```learned``` mode. I'd suggest at first not to set ```--brevity-penalty-constant-length-ratio``` with the ```constant``` mode and see what happens. This will use length ratio as estimated on the training set. If that doesn't work, change the constant mode to the ```learned``` one, it's more powerful in that it adapts to input sentences. If that doesn't help, try modifying the penalty weight parameter.

In summary parameters that can encourage longer outputs:
- ```--brevity-penalty-type``` in the ```constant``` mode:
  - ```--brevity-penalty-weight``` (the higher the longer)
  - ```--brevity-penalty-constant-length-ratio``` (the higher the longer)
- ```--brevity-penalty-type``` in the ```learned``` mode:
  - ```--brevity-penalty-weight``` (the higher the longer)

Also, try looking at the global length ratio -- some individual examples may not change but overall length can go up on average.

**Comment by artemsok** [comment]
Closing for now, feel free to reopen if questions remain
